core/common/src/LocalDateRange.kt

```diff
@@ -0,0 +1,263 @@
+ */
+public fun LocalDateProgression.first(): LocalDate {
+    if (isEmpty())
+        throw NoSuchElementException("Progression $this is empty.")
+    return this.first
+}
+
+/**
+ * Returns the last [LocalDate] of the [LocalDateProgression].
+ *
+ * @throws NoSuchElementException if the progression is empty.
+ *
+ * @sample kotlinx.datetime.test.samples.LocalDateRangeSamples.firstAndLast
+ */
+public fun LocalDateProgression.last(): LocalDate {
+    if (isEmpty())
+        throw NoSuchElementException("Progression $this is empty.")
+    return this.last
+}
+
+/**
+ * Returns the first [LocalDate] of the [LocalDateProgression], or null if the progression is empty.
+ *
+ * @sample kotlinx.datetime.test.samples.LocalDateRangeSamples.firstAndLast
+ */
+public fun LocalDateProgression.firstOrNull(): LocalDate? = if (isEmpty()) null else this.first
+
+/**
+ * Returns the last [LocalDate] of the [LocalDateProgression], or null if the progression is empty.
+ *
+ * @sample kotlinx.datetime.test.samples.LocalDateRangeSamples.firstAndLast
+ */
+public fun LocalDateProgression.lastOrNull(): LocalDate? = if (isEmpty()) null else this.last
+
+/**
+ * Returns a reversed [LocalDateProgression], i.e. one that goes from [last] to [first].
+ * The sign of the step is switched, in order to reverse the directionality of the progression.
+ *
+ * @sample kotlinx.datetime.test.samples.LocalDateRangeSamples.reversedProgression
+ */
+public fun LocalDateProgression.reversed(): LocalDateProgression = LocalDateProgression(longProgression.reversed())
+
+/**
+ * Returns a [LocalDateProgression] with the same start and end, but a changed step value.
+ *
+ * **Pitfall**: the value parameter represents the magnitude of the step, not the direction, and therefore must be positive.
+ * Its sign will be matched to the sign of the existing step, in order to maintain the directionality of the progression.
+ * If you wish to switch the direction of the progression, use [LocalDateProgression.reversed]
+ *
+ * @sample kotlinx.datetime.test.samples.LocalDateRangeSamples.progressionWithStep
+ */
+public fun LocalDateProgression.step(value: Int, unit: DateTimeUnit.DayBased) : LocalDateProgression = step(value.toLong(), unit)
+
+/**
+ * Returns a [LocalDateProgression] with the same start and end, but a changed step value.
+ *
+ * **Pitfall**: the value parameter represents the magnitude of the step, not the direction, and therefore must be positive.
+ * Its sign will be matched to the sign of the existing step, in order to maintain the directionality of the progression.
+ * If you wish to switch the direction of the progression, use [LocalDateProgression.reversed]
+ *
+ * @sample kotlinx.datetime.test.samples.LocalDateRangeSamples.progressionWithStep
+ */
+public fun LocalDateProgression.step(value: Long, unit: DateTimeUnit.DayBased) : LocalDateProgression = LocalDateProgression(longProgression.step(safeMultiply(value, unit.days.toLong())))
```

Comment:
I don't think `step` should throw an exception if `value * unit.days` overflows a `Long`. We can just clamp the result to `Long.(MAX|MIN)_VALUE` depending on the signs of the inputs, and `step` will behave correctly.

Or even if we decide to throw a `DateTimeArithmeticException`, we must at least declare it.

In any case, it's also a thing worth an additional test.